### ISSUE TITLE
trackInAppPurchases crashes for Subscriptions

### ISSUE THREAD

**Issue by ninjudd** [segmentio/analytics-ios#666]
There is an exception in this code in `SEGStoreKitTracker.m` when in-app purchases include Subscriptions:

```
- (void)trackTransaction:(SKPaymentTransaction *)transaction forProduct:(SKProduct *)product
{
    NSString *currency = [product.priceLocale objectForKey:NSLocaleCurrencyCode];

    [self.analytics track:@"Order Completed" properties:@{
        @"orderId" : transaction.transactionIdentifier,
        @"affiliation" : @"App Store",
        @"currency" : currency,
        @"products" : @[
            @{
               @"productId" : product.productIdentifier,
               @"quantity" : @(transaction.payment.quantity),
               @"sku" : transaction.transactionIdentifier,
               @"price" : product.price,
               @"name" : product.localizedTitle
            }
        ]
    }];
}
```

Looks like `transaction.payment.quantity` is `nil` for Subscriptions.

**Comment by f2prateek** [comment]
@ninjudd I see - is it documented anywhere it can be nil? Not sure I see it https://cloudup.com/cHnuNFz2rDu

**Comment by ninjudd** [comment]
@f2prateek I dug in further, and I was wrong. The entire `transaction` is actually `nil`. Not sure why that is the case. Could be related to Subscriptions, or it could be related to being in the Sandbox environment...

**Comment by f2prateek** [comment]
Gotcha, maybe this method is being called https://github.com/segmentio/analytics-ios/blob/master/Analytics/Classes/Internal/SEGStoreKitTracker.m#L62 without this being called https://github.com/segmentio/analytics-ios/blob/master/Analytics/Classes/Internal/SEGStoreKitTracker.m#L48? This could cause the transaction for the identifier to be nil. I'm unsure under what conditions this could happen in practice however.

**Comment by f2prateek** [comment]
Tracking internally in JIRA as well — https://segment.atlassian.net/browse/LIB-130

**Comment by dlahyani** [comment]
The only way I found this can happen is if 2 transactions of the same product are reported by the `SKPaymentQueue` in a single call (or adjacent calls) to the observer's (`SEGStoreKitTracker`) `paymentQueue:updatedTransactions:` with status `SKTransacationStatePurchased` (Later on how can this happen). Then in response to each of these transactions the `SEGStoreKitTracker` adds the transaction and a new `SKProductRequest` instance to its `transactions` and `productRequests` dictionaries respectively. The problem is that `SEGStoreKitTracker` writes to the dictionaries twice with using the same key - the product ID **so basically it adds only 1 entry to the dictionary**, while it creates and starts 2 `SKProductRequest` that later (when completed) causes 2 pulls from the same dictionary using the same key. When the first `SKProductRequest` completes it actually pulls (and removes) a transaction from the dictionary and everything works as expected, but when the 2nd request completes it tries to pull the transaction but gets `nil` instead and there you have it.

Now for the $1M question - why would `SKPaymentQueue` invoke the observer's `paymentQueue:updatedTransactions:` with 2 transactions having the same productId in a single call?
The answer IMO is related to the unique behavior of subscriptions in Sandbox environment. In Sandbox  subscription period is shrunk (1 month == 5 minutes in Sandbox), so subscriptions may renew quite quickly, but that's not fast enough right? However if you purchase a monthly subscription in sandbox environment, leave the app in the background for like 10 minutes and then go back to the app you **may** find 2 transactions waiting for you in the payment queue, one for each subscription renewal. In this case `SKPaymentQueue` will likely let the observers know about these 2 transactions in a single call and there you have it.

Disclaimer: I don't user Segment Analytics framework yet (maybe in the future) so I can't really verify what I'm saying is 100% true, but I worked a lot with StoreKit and experienced few related issues myself. I suggest someone using Segment or the issue opener @ninjudd will verify my theory.

Hope this helps, and goodluck :)
